sgrep.py

```diff
@@ -229,77 +257,140 @@ def _evaluate_single_expression(
                 if is_enclosed:
                     output_ranges.add(arange)
                     break  # found a match, no need to keep going
-        # print(f"after filter `{operator}`: {output_ranges}")
+        debug_print(f"after filter `{expression.operator}`: {output_ranges}")
         return output_ranges
-    elif operator == OPERATORS.AND_NOT_INSIDE:
+    elif expression.operator == OPERATORS.AND_NOT_INSIDE:
         # remove all ranges enclosed by or equal to
         output_ranges = ranges_left.copy()
         for arange in ranges_left:
             for keep_inside_this_range in results_for_pattern:
                 if keep_inside_this_range.is_enclosing_or_eq(arange):
                     output_ranges.remove(arange)
                     break
-        # print(f"after filter `{operator}`: {output_ranges}")
+        debug_print(f"after filter `{expression.operator}`: {output_ranges}")
+        return output_ranges
+    elif expression.operator == OPERATORS.WHERE_PYTHON:
+        if not RCE_RULE_FLAG not in flags:
+            print_error_exit(
+                f"at least one rule needs to execute arbitrary code; this is dangerous! if you want to continue, enable the flag: RCE_RULE_FLAG"
+            )
+        assert expression.operand, "must have operand for this operator type"
+
+        output_ranges = set()
+        # Look through every range that hasn't been filtered yet
+        for sgrep_range in list(flatten(results.values())):
+            # Only need to check where-python clause if the range hasn't already been filtered
+
+            if sgrep_range.range in ranges_left:
+                debug_print(
+                    f"WHERE is {expression.operand}, metavars: {sgrep_range.metavars}"
+                )
+                if where_python_statement_matches(
+                    expression.operand, sgrep_range.metavars
+                ):
+                    output_ranges.add(sgrep_range.range)
+        debug_print(f"after filter `{expression.operator}`: {output_ranges}")
         return output_ranges
+
     else:
         raise NotImplementedError(
-            f"{PLEASE_FILE_ISSUE_TEXT}: unknown operator {operator}"
+            f"{PLEASE_FILE_ISSUE_TEXT}: unknown operator {expression.operator}"
         )
 
 
-def evaluate_expression(expression, results: Dict[str, List[Range]]) -> Set[Range]:
-    ranges_left = set(flatten(results.values()))
-    return _evaluate_expression(expression, results, ranges_left)
+# Given a `where-python` expression as a string and currently matched metavars,
+# return whether the expression matches as a boolean
+def where_python_statement_matches(
+    where_expression: str, metavars: Dict[str, str]
+) -> bool:
+    # TODO: filter out obvious dangerous things here
+    global output
+    output = None
+
+    # HACK: we're executing arbitrary Python in the where-python,
+    # be careful my friend
+    vars = metavars
+    try:
+        exec(f"global output; output = {where_expression}")
```

Comment:
woohoo

Reply:
What is the difference between exec and eval?

Reply:
https://stackoverflow.com/questions/2220699/whats-the-difference-between-eval-exec-and-compile/29456463#29456463 @aryx this is a great explanation